### ISSUE TITLE
Do you mind if I write a request to be added to F-Droid?

### ISSUE THREAD

**Issue by Artem13327** [kaajjo/Libre-Sudoku#7]

**Comment by kaajjo** [comment]
Hi, I don't mind, that would be great

**Comment by IzzySoft** [comment]
Done. @kaajjo once your app has entered the F-Droid.org repo, do you want me to keep it in mine? Else I'd give it a 2 weeks overlap and then remove it (default action unless explicitly requested otherwise – or I want to keep it myself for some reason :zany_face:).

**Comment by kaajjo** [comment]
@IzzySoft If it does not cause any confusion to the end users, you can keep it in your repository

**Comment by kaajjo** [comment]
@IzzySoft Thanks, all data is ok

**Comment by shuvashish76** [comment]
https://gitlab.com/fdroid/rfp/-/issues/2315

**Comment by IzzySoft** [comment]
@kaajjo as nobody had picked it up, I gave it a try myself now. Unfortunately the app fails to build:

```
WARNING: /builds/fdroid/fdroiddata/build/com.kaajjo.libresudoku/app/build/intermediates/merged_java_res/release/base.jar: R8: Unexpected reference to missing service class: META-INF/services/javax.annotation.processing.Processor.
ERROR: Missing classes detected while running R8. Please add the missing classes or apply additional keep rules that are generated in /builds/fdroid/fdroiddata/build/com.kaajjo.libresudoku/app/build/outputs/mapping/release/missing_rules.txt.
ERROR: R8: Missing class javax.annotation.processing.AbstractProcessor (referenced from: void com.google.auto.service.processor.AutoServiceProcessor.<init>() and 1 other context)
Missing class javax.annotation.processing.SupportedOptions (referenced from: com.google.auto.service.processor.AutoServiceProcessor)
```

Any idea what might cause that? You can find my merge request [here](https://gitlab.com/fdroid/fdroiddata/-/merge_requests/12992), and [this](https://gitlab.com/fdroid/fdroiddata/-/jobs/4194109286) would be the corresponding pipeline's log.

**Comment by kaajjo** [comment]
@IzzySoft I thought it was my local android studio error, it just randomly showed up when I updated the libraries. I will push the fix now. Thank you

**Comment by IzzySoft** [comment]
Thanks! Please give a ping when we can try again (and which commit we should take – unless you create a new release, in which case it's pretty clear then :smile:).

**Comment by IzzySoft** [comment]
PS: I just see there's a commit that sounds like the fix. Will try it now.

**Comment by IzzySoft** [comment]
Thanks, runs through now (build [still fails](https://gitlab.com/fdroid/fdroiddata/-/jobs/4195346203) but that's my fault: need to find out how to specify in the YAML that per-ABI APKs are expected). Will ask for help from my team mates (I'm not an Android dev so i lack some experience in that field :see_no_evil:).

We can check if the build is reproducible if you'd provide us the APKs built from the same commit and signed by you. If that succeeds, we'd ship yours then – so cross-updates would be possible. Maybe you consider that while I see to fix the build recipe?

**Comment by IzzySoft** [comment]
OK, I've disabled per-ABI splits now (APK size is quite small) as we cannot have multiple APKs with the same `versionCode`. So for testing reproducible builds, I'd need the "universal APK" then – should you wish to go that path.

**Comment by kaajjo** [comment]
Here is the universal apk signed by me
[libresudoku-universal-release.zip](https://github.com/kaajjo/Libre-Sudoku/files/11351826/libresudoku-universal-release.zip)

**Comment by IzzySoft** [comment]
Thanks! Not that far from being reproducible – but still:

```
--- /dev/fd/63	2023-04-28 11:23:58.182298868 +0200
+++ /dev/fd/62	2023-04-28 11:23:58.182298868 +0200
@@ -567,7 +567,7 @@
   res/iw.xml
   32-bit CRC value (hex):                         5b72ebd5
   res/iz.json
-  32-bit CRC value (hex):                         1ebffb6c
+  32-bit CRC value (hex):                         50300015
   res/j-.9.png
   32-bit CRC value (hex):                         d42327b8
   res/j9.9.png
@@ -911,19 +911,19 @@
   META-INF/kotlinx_coroutines_core.version
   32-bit CRC value (hex):                         10dbda9d
   META-INF/services/javax.annotation.processing.Processor
-  32-bit CRC value (hex):                         41770213
+  32-bit CRC value (hex):                         e2769624
   META-INF/services/kotlinx.coroutines.CoroutineExceptionHandler
-  32-bit CRC value (hex):                         167298eb
+  32-bit CRC value (hex):                         8f594b50
   META-INF/services/kotlinx.coroutines.internal.MainDispatcherFactory
-  32-bit CRC value (hex):                         bd6769a0
+  32-bit CRC value (hex):                         4b95597f
   META-INF/services/org.acra.collector.Collector
-  32-bit CRC value (hex):                         49261d88
+  32-bit CRC value (hex):                         ea9f59be
   META-INF/services/org.acra.interaction.ReportInteraction
-  32-bit CRC value (hex):                         a99ed097
+  32-bit CRC value (hex):                         b3702163
   META-INF/services/org.acra.sender.ReportSenderFactory
-  32-bit CRC value (hex):                         a210a3e9
+  32-bit CRC value (hex):                         e6dc7eac
   META-INF/services/org.acra.startup.StartupProcessor
-  32-bit CRC value (hex):                         8fa4c296
+  32-bit CRC value (hex):                         50528c53
   kotlin-tooling-metadata.json
   32-bit CRC value (hex):                         fdcc39d3
   kotlin/annotation/annotation.kotlin_builtins
```

Not sure what it is. Here's the full diff, if you could have a look: [apkdiff.zip](https://github.com/kaajjo/Libre-Sudoku/files/11352092/apkdiff.zip) Ideas of possible causes:

* Java version (ours was built using Java 17)
* line endings (are you building on Windows?)
* Android Studio: did you build from within AS? If so, could you try building from the command line (and I'd check again with that)?

**Comment by kaajjo** [comment]
I tried to build on linux
JDK version is 17.0.6, provided by android studio
[app-release.zip](https://github.com/kaajjo/Libre-Sudoku/files/11352568/app-release.zip)

**Comment by IzzySoft** [comment]
looks much better, thanks!

```
--- /dev/fd/63	2023-04-28 13:14:33.954488240 +0200
+++ /dev/fd/62	2023-04-28 13:14:33.954488240 +0200
@@ -567,7 +567,7 @@
   res/iw.xml
   32-bit CRC value (hex):                         5b72ebd5
   res/iz.json
-  32-bit CRC value (hex):                         b9fdab1d
+  32-bit CRC value (hex):                         50300015
   res/j-.9.png
   32-bit CRC value (hex):                         d42327b8
   res/j9.9.png
```

What's that JSON file? Looks like that's generated at build time, and thus the timestamp (plus maybe the contents) are an issue:

```json
{"metadata":{"generated":"2023-04-28T07:51:31.228Z"},"libraries":[{"uniqueId":"ch.acra:acra-dialog"
```

Maybe that can be kept out of the APK? It's the only thing blocking reproducible builds now it seems.

**Comment by kaajjo** [comment]
I think it's related to [AboutLibraries](https://github.com/mikepenz/AboutLibraries) library. Here is the issue https://github.com/mikepenz/AboutLibraries/issues/784
I added changes to exclude generated fields, see if it helps 
[app-release.zip](https://github.com/kaajjo/Libre-Sudoku/files/11353319/app-release.zip)

**Comment by IzzySoft** [comment]
Ah, yes, that rings a bell! So I need to trigger another build on our end first, based on the very same commit. I cannot find a new one; did you push it already?

**Comment by kaajjo** [comment]
Oops, pushed now 6ac7d542b8fb4261a080becac0b3127d2529317c

**Comment by IzzySoft** [comment]
Build running. I guess that, seeing the APK file size, you'll stick with the universal build? I see that this commit also disables splits, so I assume it will stay (and thus removed it from our recipe to avoid it to "err out" if it found nothing to replace). Should you plan to put it back, we'd need to adjust on our end as well.

**Comment by kaajjo** [comment]
Yes, I will stick with the universal build

**Comment by IzzySoft** [comment]
Yay, and congrats! I hereby confirm a reproducible build :partying_face:

Can you please attach that last APK to the 1.5.0 release – and do so for future releases as well? Just keep the name as-is, the tag name will take care then:

    https://github.com/kaajjo/Libre-Sudoku/releases/download/v%v/app-release.apk

will become

    https://github.com/kaajjo/Libre-Sudoku/releases/download/v1.5.0/app-release.apk

after `%v` was replaced with the `versionName`. I'd then add

```yaml
Binaries: https://github.com/kaajjo/Libre-Sudoku/releases/download/v%v/app-release.apk
```

to the YAML. Then, whenever build and RB check succeed, **your** signed APK will be distributed.

**Comment by kaajjo** [comment]
Very good, thank you so much ❤️
I added the apk to the 1.5.0 release

**Comment by IzzySoft** [comment]
Updated the YAML. Prepare the drum roll! :crossed_fingers:

**Comment by IzzySoft** [comment]
Yay, congrats! All the CI lights are turning green :smiley: Marked the MR for review then. Afraid I won't find the time to perform the review myself (will not be able to work on my computer for the next 10 days). Should the review be still pending after that, feel free to give me a ping (though I'm afraid it will then take me 2 days before I got through the mail queue to reach your notification :see_no_evil:)

**Comment by IzzySoft** [comment]
Oh my :see_no_evil: Ever tried it on a smaller screen? You'll get stuck in the onboarding. Because when trying to tap the "start" button, you're back at your homescreen. Having tricked it, the icon size is a bit… weird?

[<img src="https://user-images.githubusercontent.com/6781438/235249660-bcf7d64d-3ece-4405-a993-83f69e7ad8ee.png" width="200">](https://user-images.githubusercontent.com/6781438/235249660-bcf7d64d-3ece-4405-a993-83f69e7ad8ee.png) [<img src="https://user-images.githubusercontent.com/6781438/235250167-45727853-7674-4bb3-96a4-61d15045fa93.png" width="200">](https://user-images.githubusercontent.com/6781438/235250167-45727853-7674-4bb3-96a4-61d15045fa93.png)

My review is complete, your app will show up at F-Droid in a few days – welcome aboard! Those quirks are something to be fixed with the next release :wink: That said: Do you want me to keep your app in my repo? I usually remove an app after about 2 weeks of its reaching F-Droid.org – but if requested by its author, it can also be kept. Thanks to reproducible builds that would mean faster updates (my updater picks up within 24h, F-Droid's needs to build first and thus takes 3..5 days. Nothing crucial here, but I leave the decision to you.

**Comment by kaajjo** [comment]
> Ever tried it on a smaller screen?

I have tried, but not on that small. The reason for the weird size of the icons at the bottom of the screen is also the very small screen. Will be fixed 😅

> Do you want me to keep your app in my repo?

Let's keep it

Thanks for your work and for liking my app 😄

**Comment by IzzySoft** [comment]
Thanks! Thought so (screen size), and also thought so (will be fixed) – which is why I merged it right away: it's not a show stopper. And that it glued me to the screen for an hour solving one Sudoku (it's quite a while ago I did that last, and I really like it – which is why I never installed a Sudoku app on my device, and had to uninstall yours right after the review :see_no_evil: Am much tempted to reinstall…)

> Let's keep it

Yes, will do! Just wanted to mark it for keep and noticed I've asked that before, LOL… Now I know why I named that feature "DoubleList" in my config: I ask twice for the listing :zany_face:

> Thanks for your work and for liking my app

Gladly done! And that line at the bottom (where the icons currently are too small) were what made me especially liking it. I vaguely remember having looked at another Sudoku app and had trouble with "fill versus note". That's implemented very nicely with yours! Just hit the pen for a note, hit it again (uncheck) for the shoot.

So I guess this issue is solved now and can be closed – at least once LibreSudoku shows up at F-Droid in a few days. Let me check [the build monitor](https://monitor.f-droid.org/builds/running)… OK, should be picked up with the next round, which probably starts tomorrow – which means it would be listed in the first half of next week if all goes to plan.

Thanks for providing such a nice app the FOSS way, @kaajjo! :heart_eyes:
